File: src/Solutions/Euler018.hs
module Solutions.Euler018 where
import Utils.Input(getInput)

calcTwoRows :: [Int] -> [Int] -> [Int]
calcTwoRows row1 row2 =
    let __c _ [] acc = reverse acc
        __c (a:b:t1) (c:t2) acc = __c (b:t1) t2 ((max a b + c):acc)
    in __c row1 row2 []

calcMax :: [[Int]] -> Int
calcMax xss = case xss of
    [[a]] -> a
    h1:h2:t -> calcMax $ calcTwoRows h1 h2:t

process = calcMax . reverse . map (map read . words) . lines

euler018 :: IO String
euler018 = show . process <$> getInput 18

File: rust/src/solutions/euler018.rs
use crate::utils::input::get_input;

fn parse_line(line: &str) -> Vec<i32> {
    line.split_whitespace()
        .map(|w| w.parse().unwrap())
        .collect()
}

fn parse_input(num: usize) -> Vec<Vec<i32>> {
    get_input(num).lines().map(parse_line).collect()
}

fn find_optimal_path(mut triangle: Vec<Vec<i32>>) -> i32 {
    for i in (0..triangle.len() - 1).rev() {
        for j in 0..=i {
            triangle[i][j] += std::cmp::max(triangle[i + 1][j], triangle[i + 1][j + 1]);
        }
    }
    triangle[0][0]
}

pub fn solve_triangle(num: usize) -> String {
    find_optimal_path(parse_input(num)).to_string()
}

pub fn euler018() -> String {
    solve_triangle(18)
}
